watchtower/wtclient/client.go

```diff
@@ -0,0 +1,810 @@
+			return s.ForceQuit
+		})
+
+		log.Infof("Watchtower client unclean shutdown complete, "+
+			"stats: %s", c.stats)
+	})
+}
+
+// RegisterChannel persistently initializes any channel-dependent parameters
+// within the client. This should be called during link startup to ensure that
+// the client is able to support the link during operation.
+func (c *TowerClient) RegisterChannel(chanID lnwire.ChannelID) error {
+	c.sweepPkScriptMu.Lock()
+	defer c.sweepPkScriptMu.Unlock()
+
+	// If a pkscript for this channel already exists, the channel has been
+	// previously registered.
+	if _, ok := c.sweepPkScripts[chanID]; ok {
+		return nil
+	}
+
+	// Otherwise, generate a new sweep pkscript used to sweep funds for this
+	// channel.
+	pkScript, err := c.cfg.NewAddress()
+	if err != nil {
+		return err
+	}
+
+	// Persist the sweep pkscript so that restarts will not introduce
+	// address inflation when the channel is reregistered after a restart.
+	err = c.cfg.DB.AddChanPkScript(chanID, pkScript)
+	if err != nil {
+		return err
+	}
+
+	// Finally, cache the pkscript in our in-memory cache to avoid db
+	// lookups for the remainder of the daemon's execution.
+	c.sweepPkScripts[chanID] = pkScript
+
+	return nil
+}
+
+// BackupState initiates a request to back up a particular revoked state. If the
+// method returns nil, the backup is guaranteed to be successful unless the:
+//  - client is force quit,
+//  - justice transaction would create dust outputs when trying to abide by the
+//    negotiated policy, or
+//  - breached outputs contain too little value to sweep at the target sweep fee
+//    rate.
+func (c *TowerClient) BackupState(chanID *lnwire.ChannelID,
+	breachInfo *lnwallet.BreachRetribution) error {
+
+	// Retrieve the cached sweep pkscript used for this channel.
+	c.sweepPkScriptMu.RLock()
+	sweepPkScript, ok := c.sweepPkScripts[*chanID]
+	c.sweepPkScriptMu.RUnlock()
+	if !ok {
+		return ErrUnregisteredChannel
+	}
+
+	task := newBackupTask(chanID, breachInfo, sweepPkScript)
+
+	return c.pipeline.QueueBackupTask(task)
+}
+
+// nextSessionQueue attempts to fetch an active session from our set of
+// candidate sessions. Candidate sessions with a differing policy from the
+// active client's advertised policy will be ignored, but may be resumed if the
+// client is restarted with a matching policy. If no candidates were found, nil
+// is returned to signal that we need to request a new policy.
+func (c *TowerClient) nextSessionQueue() *sessionQueue {
+	// Select any candidate session at random, and remove it from the set of
+	// candidate sessions.
+	var candidateSession *wtdb.ClientSession
+	for id, sessionInfo := range c.candidateSessions {
+		delete(c.candidateSessions, id)
+
+		// Skip any sessions with policies that don't match the current
+		// configuration. These can be used again if the client changes
+		// their configuration back.
+		if sessionInfo.Policy != c.cfg.Policy {
```

Comment:
What ends up happening with these sessions on the server side? Are they no longer being watched?

Reply:
the server is always watching all updates it has ever received. this check just prevents against the case where we used half of a session and restart with different parameters. we skip the old policies so that if we ever restart with the prior params, we can resume using that session